Add mypy as a flycheck checker

Wilfred opened the pull request:
This is based on https://github.com/lbolla/emacs-flycheck-mypy and
adds support for warnings too.

This was originally proposed by Lunaryorn: https://github.com/lbolla/emacs-flycheck-mypy/issues/1 and the original author is happy with us doing this.

Mypy is becoming more common in the Python world, it doesn't overlap with pylint or flake8, so I think it's a great fit for flycheck.

Let me know if I've missed anything or you'd like any changes.

cpitclaudel commented on flycheck.el:
```diff
@@ -8788,6 +8789,26 @@ See URL `https://docs.python.org/3.4/library/py_compile.html'."
           line ", " column ", " (one-or-more not-newline) line-end))
   :modes python-mode)
 
+(flycheck-def-args-var flycheck-python-mypy-args python-mypy)
```
We usually avoid generic `-args` variables, in favor of specific variables for each relevant flag; I think it's be better to do this here too.

Wilfred replied:
Which do you think are the most relevant?

```
$ mypy --help
usage: mypy [-h] [-v] [-V] [--python-version x.y] [--platform PLATFORM] [-2]
            [--ignore-missing-imports]
            [--follow-imports {normal,silent,skip,error}]
            [--disallow-any {unimported, expr, unannotated, decorated, explicit, generics}]
            [--disallow-untyped-calls] [--disallow-untyped-defs]
            [--disallow-incomplete-defs] [--check-untyped-defs]
            [--disallow-subclassing-any] [--warn-incomplete-stub]
            [--disallow-untyped-decorators] [--warn-redundant-casts]
            [--no-warn-no-return] [--warn-return-any] [--warn-unused-ignores]
            [--warn-unused-configs] [--show-error-context]
            [--no-implicit-optional] [-i] [--quick-and-dirty]
            [--cache-dir DIR] [--skip-version-check] [--strict-optional]
            [--strict-optional-whitelist [GLOB [GLOB ...]]]
            [--junit-xml JUNIT_XML] [--pdb] [--show-traceback] [--stats]
            [--inferstats] [--custom-typing MODULE]
            [--custom-typeshed-dir DIR] [--scripts-are-modules]
            [--config-file CONFIG_FILE] [--show-column-numbers]
            [--find-occurrences CLASS.MEMBER] [--strict]
            [--shadow-file SOURCE_FILE SHADOW_FILE] [--any-exprs-report DIR]
            [--cobertura-xml-report DIR] [--html-report DIR]
            [--linecount-report DIR] [--linecoverage-report DIR]
            [--memory-xml-report DIR] [--old-html-report DIR]
            [--txt-report DIR] [--xml-report DIR] [--xslt-html-report DIR]
            [--xslt-txt-report DIR] [-m MODULE] [-c PROGRAM_TEXT] [-p PACKAGE]
            [files [files ...]]
```

I like to run with `--py2 --strict --disallow-any=generics` for new scripts, or `--py2 --check-untyped-defs` when I'm gradually porting a codebase (to check calls to annotated functions even when the caller isn't yet annotated).

`--strict` includes quite a few arguments -- do we offer the individual arguments too?

cpitclaudel replied:
My usual approach is to add whatever the original writer of the checker finds useful, and accept pull requests for more as needed :)

cpitclaudel commented on flycheck.el:
```diff
@@ -8788,6 +8789,26 @@ See URL `https://docs.python.org/3.4/library/py_compile.html'."
+Customize `flycheck-python-mypy-args` to add specific args to default
+executable.
+
+E.g. when processing Python 2 files, add \"--py2\".
```
For example, this could controled by a `flycheck-mypy-python-version` variable.

cpitclaudel commented on flycheck.el:
```diff
@@ -8788,6 +8789,26 @@ See URL `https://docs.python.org/3.4/library/py_compile.html'."
+
+  :command ("mypy"
+            (eval flycheck-python-mypy-args)
+            source-original)
```
mypy supports stdin, so we should use that.

Wilfred replied:
How do you feel about using `source-inplace` here instead? That way mypy can follow imports too.

cpitclaudel replied:
mypy can follow imports if you give it --shadow-file, and this way we don't need to pollute the current folder with `flycheck_`-prefixed files

Wilfred commented:
@cpitclaudel thanks for sharing your version!

I tend to use mypy with flake8 afterwards, so what should I put as `:next-checkers`? I see you've used `python-pylint`, but that doesn't seem to have `:next-checkers` configured.

dakra commented:
I use https://github.com/flycheck/flycheck/pull/1080 since 10 month now. Like in the comment there, I had to rename some arguments that got changed in newer mypy versions but otherwise works like a charm.

So I'm +1 for merging this here or #1080 or any mypy checker really ;)

cpitclaudel commented:
> I tend to use mypy with flake8 afterwards, so what should I put as :next-checkers? I see you've used python-pylint, but that doesn't seem to have :next-checkers configured.

It's an annoying issue, indeed, which is what stalled my previous PR for a mypy checker.
I'm OK with leaving flake8 in there.

dakra commented:
@Wilfred bump
I would like a mypy checker :)

Wilfred commented:
@cpitclaudel could you review this again? I think this is in much better shape now.

We now have just one configurable option, the path to a mypy.ini configuration file. This allows you to configure mypy in any way, which I find really convenient. If you have different mypy.ini files you can use .dir-locals.el.

This checker works pretty well for buffers that are backed by files, but I don't know how to handle other buffers. I want to do something like this:

```emacs-lisp
(if source-original
    ;; If this file is backed by a buffer, tell mypy where the
    ;; original file was but shadow it with a temporary file of the
    ;; current buffer contents.
    (concat "--shadow-file" source-original source source-original)
  ;; Just run mypy on a temporary file containing the buffer contents.
  source)
```

but I'm not sure how to express this with `eval` or the `:command` DSL.

Wilfred commented:
Added docs and changed the checker to use `(config-file ...)` which I hadn't spotted before.

fmdkdd commented:
> This checker works pretty well for buffers that are backed by files, but I don't know how to handle other buffers.

If `--shadow-file` is indeed checking the temporary created by `source`, then it should work seemlessly for buffers not backed by files.  Is this not the case?

Wilfred commented:
@fmdkdd `--shadow-file` allows you to say "use this temporary file as if it was this other file on disk". There is no 'other file on disk' for buffers that aren't backed by files.

fmdkdd commented:
> There is no 'other file on disk' for buffers that aren't backed by files.

Indeed, but why does `mypy` care about the actual file, since we already gave it the content it should check?  My question is, if we pass `foo.py` as a phony file path, with a real temporary file given to `--shadow-file`, does `mypy` complain?

cpitclaudel commented:
@fmdkdd I think it's a mypy bug:

```
$ cd /tmp/
$ mkdir python
$ cd python/
$ echo '"""test"""' > a.py
$ mypy --shadow-file b.py a.py a.py
mypy: can't read file 'a.py': No such file or directory
```

cpitclaudel commented:
I opened https://github.com/python/mypy/issues/4746

fmdkdd commented:
Great!  Keep us posted.

cpitclaudel commented:
Edit: acknowledged as an issue, but I don't have time to fix it myself, so let's just add a check to ensure that the buffer is saved.

We should also pass a flag to prevent all these `.mypy` cache directories from being created.

Then I'd be in favor of merging.  I think it shouldn't be enabled for all files by default, though, since it's pretty slow.  Or maybe it should just run last?

fmdkdd commented:
> Edit: acknowledged as an issue, but I don't have time to fix it myself, so let's just add a check to ensure that the buffer is saved.

Agreed.  We should add a comment stating that checking on saved buffers is a workaround for the mypy bug, and add a link to it.

> We should also pass a flag to prevent all these .mypy cache directories from being created.
Then I'd be in favor of merging. I think it shouldn't be enabled for all files by default, though, since it's pretty slow. Or maybe it should just run last?

Aren't these two related?  If you leave the cache, then potentially it can be reused for later calls?

In any case, if we put `python-mypy` after `python-pycompile`, then we force users to opt-in by selecting `mypy` in their own config (since `pycompile` should always be enabled if `python` is available).  I'd rather have that than a checker disabled by default.

fmdkdd commented:
> We could add an option.

Sure.

@Wilfred Are you still free to work on this PR?  The following changes need to be made:

1. Ensure that the buffer is saved in `:predicate` to workaround python/mypy#4746 (and add a comment to that effect with a link to the bug)
2. Move `python-mypy` in `flycheck-checkers` after `python-pycompile`
3. Add a user option to disable creating `.mypy-cache` folders

Wilfred commented:
@fmdkdd thanks, I had some time today and your explicit steps were very helpful :)

I've made the changes you requested, and rebased on latest master.

fmdkdd commented on flycheck.el:
```diff
@@ -8950,6 +8951,35 @@ See URL `https://docs.python.org/3.4/library/py_compile.html'."
+.mypy_cache files."
+  :safe #'booleanp
+  :type 'boolean
+  :package-version '(flycheck . "0.16"))
```
The version number should be `32`, not `0.16`.

Wilfred replied:
Fixed.

cpitclaudel commented on doc/languages.rst:
```diff
@@ -900,6 +900,17 @@ to view the docstring of the syntax checker.  Likewise, you may use
+      .. defcustom:: flycheck-python-mypy-disable-incremental
+
+         Switches off incremental type checking. This is slower, but
+         prevents mypy from writing ``.mypy_cache`` files.
```
I'd suggest to say `creating .mypy_cache folder` instead

cpitclaudel replied:
I also wonder if the setting should be called `enable`, and be off by default.  These folders are really annoying. Opinions?

cpitclaudel replied:
Or maybe there's a way to make mypy put all these folders in one place on the file system, and we should let users customize that?

Wilfred replied:
I've updated the doc as you suggested.

I did think about having an enable flag, but this is consistent with the mypy behaviour, which enables caching by default. Personally I don't mind the .mypy_cache directory being created, and I find mypy rather slow otherwise.

fmdkdd commented:
Thanks!  One last thing: please add a line to `CHANGES.rst` under `New syntax checkers`.

It would also be nice to add a test for the checker in `flycheck-test.el`.

Wilfred commented:
I've updated `CHANGES.rst`.

I'm not sure how to add a test. Looking at e.g. `(flycheck-ert-def-checker-test python-flake8 python nil` in flycheck-test.el then it looks like the test depends on flake8 being installed.

I can see `main/requirements.txt` is installing flake8, but that doesn't look related. If I look at another test like `(flycheck-ert-def-checker-test puppet-parser puppet parser-error-puppet-4` in flycheck-test.el then it seems to be shelling out to `puppet`, but I can't see puppet being installed either.

fmdkdd commented:
@Wilfred right, so `make integ` can be run locally and it runs on our CI.  For both, the command will run only the tests of checkers for which there is an executable present.  The tests do not install the executable.

So, if you have `mypy` installed, you can add a test, and it should run locally with `make integ`.  Then, it will be skipped on CI because there `mypy` is not installed.  For that, we need to add `mypy` to the docker image we use for running the tests, over at [flycheck/docker-tools](https://github.com/flycheck/docker-tools).

I've just added `mypy` there https://github.com/flycheck/docker-tools/commit/f32347c4d56afa9c272577cb6311215277974186, so that should be good.

For writing the test, you probably want to set the `flycheck-disabled-checkers` variable (as in the `python-plint` tests), so that Flycheck picks `python-mypy` for the buffer being tested.

Wilfred commented:
@fmdkdd that's really helpful, thanks. I've had a crack at writing a test and I think it's correct now.

I also noticed an issue with `--disable-incremental` not actually doing what I expected, so I've added a new configuration variable to disable writing `.mypy_cache` properly.

Finally, I've corrected the docs regarding the minimum mypy version required.

fmdkdd commented:
@Wilfred Thanks!  That's real nice.  Looks like your test is passing on the CI, but now we some checkers are freaking out because of we are using Python3 instead of Python2.

I'll look into that and restart the build if necessary.

fmdkdd commented:
Ok so the python-pylint failures are due to `pylint` not being found when using `python3`.   Looks like a consequence of #1113.  In the Docker image:

```
$ pylint
```
invokes `pylint` correctly, but what we actually run by default is:
```
$ python -c 'import sys,runpy;sys.path.pop(0); runpy.run_module("pylint")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python2.7/runpy.py", line 182, in run_module
    mod_name, loader, code, fname = _get_module_details(mod_name)
  File "/usr/lib/python2.7/runpy.py", line 107, in _get_module_details
    raise error(format(e))
ImportError: No module named pylint
```

Changing `python` to `python3` in the checker fixes it, but we probably don't want to do that.

Now, my understanding is that as it stands, python checkers need to be configured in order to work properly.  This was not case before #1113 if installing tools with `pip`.

Of course, one solution for the CI is to use python2 for all the Python tools, and use python3 for mypy.  But using python3 for all seemed simpler.

@cpitclaudel What is your take on this?

cpitclaudel commented:
@fmdkdd it should be enough to let-bind flycheck-python-pylint-executable to `python3` around the test — does that work?

fmdkdd commented:
> it should be enough to let-bind flycheck-python-pylint-executable to python3 around the test — does that work?

Yes it works.  And so does let-binding it to `pylint` more directly.  But what I was worried about was that users who had a similar setup would have to jump through the same hoops.  I thought the point of #1113 was to make the switch between `pylint for python2` and `pylint for python3` easier, and now I'm unclear as to how it eases that.

cpitclaudel commented:
>  But what I was worried about was that users who had a similar setup would have to jump through the same hoops. I thought the point of #1113 was to make the switch between pylint for python2 and pylint for python3 easier, and now I'm unclear as to how it eases that.

Correct (more precisely, the point of #1113 was to make it easy to use the pylint matching your version of python.  If you're using python3, you set the both python-shell-executable and flycheck-python-pylint-executable to `python3`; for python 2, you set both of them to `python2`.

(Did I get the question right?)

Wilfred commented:
Is there anything else you need from me? :)

fmdkdd commented:
@Wilfred I don't think so.  I just need to fix the python integration tests on CI and we can merge.

On a related note, has anyone taken a look at https://github.com/facebook/pyre-check/, how it compares to mypy, and whether we could integrate it with Flycheck?
